pkg/insights/insightsreport/insightsreport.go

```diff
@@ -0,0 +1,164 @@
+type InsightsReporter interface {
+	ArchiveUploaded() <-chan struct{}
+}
+
+var (
+	// ErrWaitingForVersion contains an error when the cluster version is not loaded yet
+	ErrWaitingForVersion = fmt.Errorf("waiting for the cluster version to be loaded")
+
+	// insightsStatus contains a metric with the latest report information
+	insightsStatus = metrics.NewGaugeVec(&metrics.GaugeOpts{
+		Namespace: "health",
+		Subsystem: "statuses",
+		Name:      "insights",
+		Help:      "Information about the cluster health status as detected by Insights tooling.",
+	}, []string{"metric"})
+)
+
+// New initializes and returns a Gatherer
+func New(client *insightsclient.Client, configurator Configurator, reporter InsightsReporter) *Gatherer {
+	return &Gatherer{
+		Simple:                controllerstatus.Simple{Name: "insightsreport"},
+		configurator:          configurator,
+		client:                client,
+		archiveUploadReporter: reporter.ArchiveUploaded(),
+	}
+}
+
+// PullSmartProxy performs a request to the Smart Proxy and unmarshal the response
+func (r *Gatherer) PullSmartProxy() {
+	klog.Info("Pulling report from smart-proxy")
+	ctx, cancelFunc := context.WithTimeout(context.Background(), time.Minute)
+	defer cancelFunc()
+
+	klog.V(4).Info("Retrieving report")
+	config := r.configurator.Config()
+	reportEndpoint := config.ReportEndpoint
+	reportBody, err := r.client.RecvReport(ctx, reportEndpoint)
+	if authorizer.IsAuthorizationError(err) {
+		r.Simple.UpdateStatus(controllerstatus.Summary{
+			Operation: controllerstatus.DownloadingReport,
+			Reason:    "NotAuthorized",
+			Message:   fmt.Sprintf("Auth rejected for downloading latest report: %v", err),
+		})
+		return
+	} else if err != nil {
+		klog.Errorf("Error retrieving the report: %s", err)
+		return
+	}
+
+	klog.V(4).Info("Report retrieved")
+	reportResponse := Response{}
+
+	if err = json.NewDecoder(*reportBody).Decode(&reportResponse); err != nil {
+		klog.Error("The report response cannot be parsed")
+		return
+	}
+
+	klog.V(4).Info("Smart Proxy report correctly parsed")
+
+	if r.LastReport.Meta.LastCheckedAt == reportResponse.Report.Meta.LastCheckedAt {
+		klog.V(2).Info("Retrieved report is equal to previus one. Retrying...")
+		return
+	}
+
+	defer close(r.insightsReport)
+	updateInsightsMetrics(reportResponse.Report)
+	r.LastReport = reportResponse.Report
+	return
+}
+
+// Run goroutine code for gathering the reports from Smart Proxy
+func (r *Gatherer) Run(ctx context.Context) {
+	r.Simple.UpdateStatus(controllerstatus.Summary{Healthy: true})
+	klog.V(2).Info("Starting report retriever")
+
+	for {
+		// always wait for new uploaded archive or insights-operator ends
+		select {
+		case <-r.archiveUploadReporter:
+			// When a new archive is uploaded, try to get the report and repeat every 30s
+			// until the report is retrieved
+			r.insightsReport = make(chan struct{})
+			wait.Until(r.PullSmartProxy, time.Duration(30e9), r.insightsReport)
```

Comment:
It is better to use 10*time.Second than time.Duration(30e9).

Reply:
Thank you Martin. I'm working currently on several updates for this PR and definitely it will on the boat :-)